### PR TITLE
catch Flint aborts before they happen; fixes #793

### LINKED ISSUE #793: series_flint: catch Flint aborts before they happen
```
In [2]: from symengine import *

In [3]: x=Symbol('x')                      

In [4]: series(1/cos(x/log(x)), x, 0)
Out[4]: 
         2            4   
        x          5⋅x         ⎛ 6⎞
1 + ───────── + ────────── + O⎝x ⎠
         2            4       
    2⋅log (x)   24⋅log (x)  

In [5]: series(1/cos(x/log(x)), x, 0, method='symengine')
Exception (fmpq_poly_log_series). Constant term != 1.

Abort caught. Printing stacktrace:

Traceback (most recent call last):

Done.
Aborted
```
Synopsis: Flint's C layer does not have exceptions. It simply prints a message and calls `abort()`. The only layer aware of this is Teuchos which has installed a signal handler that tries to print a stacktrace. I don't think we can throw a `RuntimeError` at that point in time.

So this must be handled by catching all conditions in SymEngine that can lead to a Flint abort before Flint is called. So we opened a SymEngine issue.

### PR DESCRIPTION
**Pull request by rwst**

### CODE REVIEW

File: symengine/series_flint.cpp [hunk 1]
```diff
@@ -2,6 +2,31 @@
...
+}
+
+void try_and_catch_abort (std::function< void ()> func)
+{
+    if (setjmp (env) == 0) {
+        signal(SIGABRT, &on_sigabrt);
+        func();
+    }
+    else {
+        std::cout << "aborted\n";
```
Comment (isuruf) [inline symengine/series_flint.cpp]: You should restore the teuchos handler here and also throw the runtime_error from here

**Comment by isuruf** [comment]
Is this thread safe?
Can `jmp_buf env;` be declared inside `try_and_catch_abort` and `on_sigabrt` be declared as a lambda referencing the `env`?

**Comment by certik** [comment]
I am not sure about this. When does flint abort?

**Comment by fredrik-johansson** [comment]
Flint aborts when given illegal input, e.g. when asked to divide by zero. It also aborts when running out of memory (which usually is due to illegal input as well, typically a negative C long being passed where a positive value is expected). I would advise against relying on being able to pass invalid input and catching the abort signal like an exception.

For example, fmpq_poly_exp_series requires that the constant term is 0, so if you have an arbitrary series, you could just check that this is the case first and use a different implementation otherwise.

I don't know if the C++ interface provides exceptions as an alternative.

**Comment by certik** [comment]
I agree with @fredrik-johansson. The idea is that we need to make sure that we are passing valid inputs into Flint, using SymEngine's asserts, that are on in Debug mode. If anything aborts, we should just let it abort. That's also important from the library perspective, since we are just a library, and we want to leave this to the user. Notice also that Teuchos abort is not enabled by default, the user needs to call `SymEngine::print_stack_on_segfault();` to enable that.

**Comment by rwst** [comment]
Both is necessary, asserts and delegation to `UPSeriesPiranha` which can handle `sin(1)` etc. I'll open new branches.

**Pull request closed without merging**